services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -3719,10 +3731,38 @@ public void initiateDataRecovery(
     ControllerClient srcFabricChildControllerClient = getFabricBuildoutControllerClient(clusterName, sourceFabric);
     ControllerClient destFabricChildControllerClient =
         getFabricBuildoutControllerClient(clusterName, destinationFabric);
-    StoreInfo storeInfo = srcFabricChildControllerClient.getStore(storeName).getStore();
-    Optional<Version> sourceVersion = storeInfo.getVersion(version);
+    StoreResponse srcStoreResponse = srcFabricChildControllerClient.getStore(storeName);
+    if (srcStoreResponse.isError()) {
+      throw new VeniceException(
+          "Error when getting store " + storeName + " from source fabric " + sourceFabric + ": "
+              + srcStoreResponse.getError());
+    }
+    StoreInfo srcStoreInfo = srcStoreResponse.getStore();
+    if (version == VERSION_ID_UNSET) {
+      version = srcStoreInfo.getCurrentVersion();
+    }
+    Optional<Version> sourceVersion = srcStoreInfo.getVersion(version);
     if (!sourceVersion.isPresent()) {
-      throw new VeniceException("Version: " + version + " does not exist in the given source fabric: " + sourceFabric);
+      throw new VeniceException(
+          "Version " + version + " does not exist in source fabric " + sourceFabric + " store " + storeName);
+    }
+    StoreResponse destStoreResponse = destFabricChildControllerClient.getStore(storeName);
+    if (destStoreResponse.isError()) {
+      throw new VeniceException(
+          "Error when getting store " + storeName + " from destination fabric " + destinationFabric + ": "
+              + destStoreResponse.getError());
+    }
+    if (whetherToCreateNewDataRecoveryVersion(destinationFabric, clusterName, destStoreResponse.getStore(), version)) {
+      Store parentStore = getStore(clusterName, storeName);
+      int newVersion = parentStore.peekNextVersion().getNumber();
+      getVeniceHelixAdmin().setStoreLargestUsedVersion(clusterName, storeName, newVersion);
```

Comment:
Even if they get the same newVersion number, only one can add the new version to the store in destination region. Because destination region child controller acquires store write lock before adding versions. Only one request can acquire the lock and add the new version. The other request will fail with version already exists exception: https://github.com/linkedin/venice/blob/2404d2387a24c048e759b47d990a0ddee842d4be/services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java#L2085

Reply:
Yeah, destination region storeLargestUsedVersion seems fine, but we could set an inconsistent number (with regard to child store's storeLargestUsedVersion) on the parent controller side because of the race? If such inconsistency is possible and unavoidable, I feel like we are supposed to use the dest store's info to peek the next version, instead of `parentStore` at L3757, aren't we?

Reply:
Talked offline. Lei thinks of a case that multiple concurrent requests can increase parent store largestUsedVersionNumber then decrease it: both request1 and request2 calculate new version as 2. Request1 sets largestUsedVersion to 2. Request3 calculate new version as 3 and sets largestUsedVersion to 3. Request2 sets largestUsedVersion back to 2.

Great catch. This would result in largestUsedVersion in parent colo smaller than child colo, which will cause problems in the store's next push.

To resolve this issue, I included "int newVersion = parentStore.peekNextVersion().getNumber()" into the store write lock. Every request will acquire a lock before it calculates the new version number and set largestUsedVersion.